### PR TITLE
Move SCF-DIIS to Python

### PR DESCRIPTION
**Pull request by JonathonMisiewicz**
## Description
In preparation for some PRs adding EDIIS and ADIIS to Psi, this PR moves the `compute_orbital_gradient` functions, which call DIIS, to the Python layer.

There are multiple things going on:
- For pybinding, it was convenient to convert the enums in `libdiis` to _strongly typed enums_. This means the enums have their own scope, which changes how other functions had to call them. Every change in occ, dfocc, dct, fisapt, and libdiis is _solely_ due to this. This occupies commit one.
- Many other functions had to be pybound. In particular, it was necessary to convert `diis_manager` into a property. Old code would call `HF.diis_manager()`, but now they call `HF.diis_manager_`. We're not bothering to deprecate the old way of calling it. I'm not aware if there's a way to still support the old syntax. It might work if we change the property name to `diis_manager`? I haven't tested this.
- It is not possible to pybind a variadic, so I had to pybind every `set_error_vector_size` and `set_vector_size` and `add_entry` type combination that Python might use. Sad, but necessary.

And with that done, all four of the `compute_orbital_gradient` functions could finally be moved to the Python layer. `compute_orbital_gradient` is kept as a virtual function, C++ side, to signal that Psi expects such a function to exist. My tests indicate that if the function is defined C++ side, it still works, so this won't break any SCF subclasses defined in plugins.

## Todos

- [x] _Lots_ of pybind-ing involving DIIS
- [x] Alert! `HF.diis_manager()` has changed to `HF.diis_manager_`
- [x] `HF.compute_orbital_gradient` moved to the Python layer. C-side `compute_orbital_gradient` is still supported, but not used by the core Psi.

## Questions
- [ ] Who is responsible for updating the Great DFOCC Branch with the `libdiis` API change?
- [ ] How do we feel about the new functions I Pybind-ed over? Anything amiss?

## Checklist
- [x] `ctest -L scf` passes

## Status
- [x] Ready for review
- [ ] Let's discuss the two questions before merging this in, please.

### CODE REVIEW

File: psi4/driver/procrouting/scf_proc/subclass_methods.py
Comment (PhillCli) [inline psi4/driver/procrouting/scf_proc/subclass_methods.py:5]: Can you add return annotations as well?

```python
def _RHF_orbital_gradient(self, save_fock: bool, max_diis_vectors: int) -> float:
```

Comment (JonathonMisiewicz) [reply psi4/driver/procrouting/scf_proc/subclass_methods.py:5]: Sure, I'll include that in the next set of commits.

Normally when somebody who isn't a core developer leaves comments on a PR, the PR is directly relevant to their research. Can I ask if there's some special reason for you commenting on this one?

File: psi4/driver/procrouting/scf_proc/subclass_methods.py
Comment (loriab) [inline psi4/driver/procrouting/scf_proc/subclass_methods.py:10]:
```suggestion
            if self.scf_type() == "DIRECT":
                storage_policy = core.DIISManager.StoragePolicy.InCore 
            else:
                storage_policy = core.DIISManager.StoragePolicy.OnDisk
```
equivalent? I usually save `getattr` for when attribute only known in a variable. But that's my readability opinion. Your choice.

File: psi4/driver/procrouting/scf_proc/subclass_methods.py
Comment (loriab) [inline psi4/driver/procrouting/scf_proc/subclass_methods.py:5]: general info to readers: the Sphinx docs can now process type annotations, so they're not added in vain. the Sphinx docs now also hate the old Parameters\n verbose : int style in-docstring type annotation, so those're worth moving up into the signature.

**Review by loriab — APPROVED** [review]
lgtm!

File: psi4/driver/procrouting/scf_proc/subclass_methods.py
Comment (JonathonMisiewicz) [reply psi4/driver/procrouting/scf_proc/subclass_methods.py:10]: Yes, equivalent. The right thing to do for readability is probably to import `from core.DIISManager import StoragePolicy, RemovalPolicy`. Then I can have a terse ternary without using `getattr`.

Comment (loriab) [reply psi4/driver/procrouting/scf_proc/subclass_methods.py:10]: that works. only other consideration is if you'd ever want to track which paths were tested with codecov.

File: psi4/driver/procrouting/scf_proc/subclass_methods.py
Comment (PhillCli) [reply psi4/driver/procrouting/scf_proc/subclass_methods.py]: @JonathonMisiewicz not particularly. I'm using some of the python SCF API lately, glad to see more logic available at the python level. In general type hinting FTW!

Comment (PeterKraus) [reply psi4/driver/procrouting/scf_proc/subclass_methods.py]: @loriab Someone ought to note that in the "developers howto", if it's not there already.

**Comment by JonathonMisiewicz** [comment]
On hold until #2296 gets approved, to see if we can reach a consensus on pybind docstring format is now.

**Comment by JonathonMisiewicz** [comment]
Force pushed to bring this branch up-to-date. Should be good to review again.

**Review by jturney — APPROVED** [review]
Looks good!

File: psi4/driver/procrouting/scf_proc/subclass_methods.py
Comment (PeterKraus) [inline psi4/driver/procrouting/scf_proc/subclass_methods.py:52]: `self.doccpi() + virpi = self.nmopi()`

File: psi4/driver/procrouting/scf_proc/subclass_methods.py
Comment (PeterKraus) [inline psi4/driver/procrouting/scf_proc/subclass_methods.py:63]: `self.Ct()`, not `Ct_`

File: psi4/driver/procrouting/scf_proc/subclass_methods.py
Comment (PeterKraus) [inline psi4/driver/procrouting/scf_proc/subclass_methods.py:88]: I'm not a terribly big fan of this triple-assignment.

**Review by PeterKraus — APPROVED** [review]
I only have a couple very minor style suggestions. I suggest @JonathonMisiewicz can decide whether to action them or merge straight away.

File: psi4/driver/procrouting/scf_proc/subclass_methods.py
Comment (JonathonMisiewicz) [reply psi4/driver/procrouting/scf_proc/subclass_methods.py:88]: It looked strange to me at first as well, but I've made my peace with it. Putting these assignments on the same line makes it obvious that these two are _intended to be_ the same function. I was a little surprised that CUHF didn't need custom logic here.

**Pull request merged**
